testing-infra: Add retrials to network intensive parts of provisioning.

adnrs96 opened the pull request:
In this PR we are adding a single retrial to any of the development environment setup part which involves intensive network interactions. These mainly include the following:
* `apt-get install`
* `apt-get dist-upgrade`
* `pip install`
We just add a conditional similar to what we have for `npm install` which based upon `subprocess.CalledProcessError` raised in python and value of `$?` in bash.
This fixes: #2034.

smarx commented:
*Automated message from Dropbox CLA bot*

@adnrs96, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented on scripts/lib/setup_venv.py:
```diff
@@ -244,7 +244,14 @@ def do_setup_virtualenv(venv_path, requirements_file, virtualenv_args):
+        run(["pip", "install", "--upgrade", "pip", "wheel"])
+        run(["pip", "install", "--no-deps", "--requirement", requirements_file])
+    except subprocess.CalledProcessError:
+        # Must be a failure due to network connection issues. Retrying...
```
`s/Must/Might/` -- it's possible that's not the cause

adnrs96 replied:
yeah. agreed.

timabbott commented on scripts/lib/setup_venv.py:
```diff
@@ -244,7 +244,14 @@ def do_setup_virtualenv(venv_path, requirements_file, virtualenv_args):
+        # Must be a failure due to network connection issues. Retrying...
+        print(WARNING + "`pip install` failed; retrying..." + ENDC)
+        run(["pip", "install", "-U", "setuptools==35.0.2"])
+        run(["pip", "install", "--upgrade", "pip", "wheel"])
```
As a follow-up, while you're thinking about this, #5158 is an issue we should look at.

adnrs96 commented on tools/lib/provision.py:
```diff
@@ -179,7 +179,11 @@ def main(options):
 
     # setup-apt-repo does an `apt-get update`
     run(["sudo", "./scripts/lib/setup-apt-repo"])
```
Should we add L181 as well to the retrial? I guess it run's a script which only installs a single package `lsb-release`.

timabbott replied:
I think for these, I'd rather just do the overall `provision` retry than adding lines to every place we do a network thing in provision; there's a lot of them!

adnrs96 commented:
I have left a comment in code about the think I was doubtful about.

adnrs96 commented:
@timabbott I have updated the PR with changes. Please take a look.
Since we were only to rerun provision by means of a new wrapper script which would be called by travis scripts, I thought it might be worth to just add a retry it in place in the two places where this was needed (the `setup-production` and `setup-backend`). This way we are saved of running into looped calls to provisioning (Which could have been taken care of by means of having variables assigned status values).

adnrs96 commented:
I guess there is no way to trigger a rerun upon a build from having some kind of travis configuration. There console is the only way to rerun a build I guess (They might have something in api to trigger rebuilds though).

timabbott commented on scripts/lib/setup_venv.py:
```diff
@@ -244,7 +244,14 @@ def do_setup_virtualenv(venv_path, requirements_file, virtualenv_args):
+        print(WARNING + "`pip install` failed; retrying..." + ENDC)
+        run(["pip", "install", "-U", "setuptools==35.0.2"])
+        run(["pip", "install", "--upgrade", "pip", "wheel"])
+        run(["pip", "install", "--no-deps", "--requirement", requirements_file])
```
I don't like the code duplication here (it will likely lead to pain later); we should extract an `install_venv_deps` function and just call it both in the try and except blocks.

adnrs96 replied:
Sure.

timabbott commented on tools/lib/provision.py:
```diff
@@ -179,7 +179,12 @@ def main(options):
+    except subprocess.CalledProcessError:
+        # Might be a failure due to network connection issues. Retrying...
+        print(WARNING + "`apt-get -y install` failed while installing dependencies; retrying..." + ENDC)
+        run(["sudo", "apt-get", "-y", "install", "--no-install-recommends"] + APT_DEPENDENCIES[codename])
```
I have the same code duplication feedback here.

timabbott commented on tools/travis/production-helper:
```diff
@@ -27,6 +27,10 @@ apt-mark hold initramfs-tools initramfs-tools-bin oracle-java8-installer oracle-
+if ! [ $? -eq 0 ]; then
+echo "\`apt-get dist-upgrade\`: Failure occured while trying to perform distribution upgrade, Retrying...";
+apt-get dist-upgrade -y $APT_OPTIONS ;
+fi
```
You should indent the inside of shell if statements 4 spaces.

timabbott commented on tools/travis/setup-backend:
```diff
@@ -4,6 +4,13 @@ set -x
+if ! [ $? -eq 0 ]; then
+echo "\`provision\`: Something went wrong with the provisioning, might be a network issue, Retrying to provision...";
+tools/provision --travis ;
+fi
```
This should be indented properly.

timabbott commented on tools/travis/setup-production:
```diff
@@ -22,6 +22,13 @@ sudo chmod a+rX /home/travis
+# once again if that fixes our problem.
+if ! [ $? -eq 0 ]; then
+echo "\`provision\`: Something went wrong with the provisioning, might be a network issue, Retrying to provision...";
+tools/provision --travis --production-travis ;
```
You don't need semicolons at the end of lines in shell

adnrs96 commented:
@timabbott I have updated the PR to address your comments. Please take a look.

timabbott commented:
@adnrs96 I merged the first two commits, after a small change to include `setup-apt-repo` in what is being tested.  @sinwar FYI since this will cause some small merge conflicts for your apt work if you've started it.

The remaining ones don't work, because we use `set -e`, which causes the shell scripts to fail immediately when an error occurs.  `tools/check-py3` does this correctly; basically, the simplest approach is to write the retry logic as e.g. `if ! tools/provision; then`, though you can also do something like `tools/provision || failed=1`.

You can iterate quickly on you shell code by doing in a short test shell script things like `false || failed=1`, using `set -e` to match how we run our main shell scripts.

After fixing this, I recommend testing them by making the script fail unconditionally so you can verify that it actually retries.

adnrs96 commented:
@timabbott I have updated this with changes. Also I added a wrapper around `setup-static-analysis` as well. I kind of did repetition of the only two `pip install` statements there since I couldn't decide what else to do. Maybe an array could have been created and then, using a `for loop` execution of the statements in array?
PS: Last commit is just a POF commit to test the retries actually occur, and only first 4 commits are for purpose of merging.
Also I do not know what is up with coveralls. Seems like failing for all my other PR's as well.

showell commented:
I merged the first three.  I think on the fourth one, it would be better stylistically to retry them individually.

adnrs96 commented:
@showell Just updated to address the comments.

showell commented:
Merged!  Thanks @adnrs96!
